src/client/components/Buttons/Button.tsx

```diff
@@ -0,0 +1,334 @@
+import React, { FC, useState, ButtonHTMLAttributes, useMemo } from 'react';
+import { Link } from 'react-router-dom';
+import styled from 'styled-components';
+import { XOR } from 'ts-xor';
+import { Spinner, Wrapper as SpinnerWrapper } from '../Loading/Spinner';
+
+/** Utility type which makes an exclusive OR of all properties in T.  */
+type oneOf<T> = { [K in keyof T]: Pick<T, K> & FixTsUnion<T, K> }[keyof T];
+type FixTsUnion<T, K extends keyof T> = { [Prop in keyof T]?: Prop extends K ? T[Prop] : never };
+
+/** Exclusive OR between T, U, and V, factoring in all properties of each type. */
+type XOR3<T, U, V> = XOR<XOR<T, U>, V>;
+
+/** Returns true if the `url` param begins with "http" or "//", signifying an external url. */
+function isAbsoluteUrl(url: string): boolean {
+	return url.toLowerCase().startsWith('http') || url.startsWith('//');
+}
+
+type ButtonBase = {
```

Comment:
my personal pref is types in separate file since this file is kinda large, but leave decision to you :)

Reply:
Do you put them in `Button.d.ts` or `Button.types.ts` or what would you recommend?

Reply:
I think `.d.ts` makes sense